src/sources/journald.rs

```diff
@@ -281,13 +259,32 @@ where
             }
         }
     }
-}
 
-#[derive(PartialEq)]
-enum LoopState {
-    NoOp,
-    AtEnd,
-    SawRecord,
+    fn seek_checkpoint(&mut self) {
+        // Retrieve the saved checkpoint, and seek forward in the journald log
+        match self.checkpointer.get() {
+            Ok(Some(cursor)) => {
+                if let Err(err) = self.journal.seek_cursor(&cursor) {
+                    error!(
+                        message = "Could not seek journald to stored cursor",
+                        error = field::display(&err)
+                    );
+                // The cursor now points to the last successfully read
+                // record, so skip past it to any newer records.
+                } else if let Some(Err(err)) = self.journal.next() {
+                    error!(
+                        message = "Could not fetch next record after seeking to cursor",
+                        error = field::display(&err)
+                    );
+                }
+            }
+            Ok(None) => {}
+            Err(err) => error!(
+                message = "Could not retrieve saved journald checkpoint",
+                error = field::display(&err)
+            ),
+        }
+    }
 }
 
 const CHECKPOINT_FILENAME: &str = "checkpoint.txt";
```

Comment:
Should we add name of the source to the file path, similarly to how it is done in `file` source?

https://github.com/timberio/vector/blob/eb0566313849002fa820d57cc15d8a9ec957b9d3/src/sources/file.rs#L133-L136

I can imagine someone creating two `journald` sources with different configurations and default `data_dir`, which would interfere.

Reply:
That's one of my use cases I was planning to try out.

Reply:
The checkpoint file goes into a directory named for the source name, which is a required unique field in the configuration. As such, it will be a unique path and not interfere with other journal readers.

Reply:
Sorry, it was my mistake. I tried the example config from the docs and saw directory `journald-source` in the Vector's `data_dir`. Then, for some reason, I've got false assumption that `journald-source` is not name of the source, but its type.

I wonder would it make sense to actually introduce two-level directory structure in `data_dir` for all components, where the first level would contain component type (for example, `journald-source`) and the second level would contain name of the component (in the example from docs it would again be `journald-source`).

For example, the user can create `file` source named `in` at first, and then change the component type to `journald`. For now it would not cause interference because of different names of data files inside `data_dir/in`, but guaranteeing that these directories are always different would allow us to not worry about possible interference in such cases in future.